builtin/union_dataset.cc

```diff
@@ -0,0 +1,381 @@
+            uint64_t hash = (*it).first;
+            ++it;
+
+            return source->getRowName(RowHash(hash));
+        }
+
+        virtual const RowName & rowName(RowName & storage) const
+        {
+            cerr << "UNIMPLEMENTED " << __FILE__ << ":" << __LINE__ << endl;
+            throw ML::Exception("Unimplemented %s : %d", __FILE__, __LINE__);
+            uint64_t hash = (*it).first;
+            return storage = source->getRowName(RowHash(hash));
+        }
+
+        const UnionDataset::Itl* source;
+        IdHashes::const_iterator it;
+
+    };
+
+    virtual vector<Path>
+    getRowNames(ssize_t start = 0, ssize_t limit = -1) const
+    {
+        // Row names are idx.rowName where idx is the index of the dataset
+        // in the union and rowName is the original rowName.
+        vector<RowName> result;
+        for (int i = 0; i < datasets.size(); ++i) {
+            const auto & d = datasets[i];
+            for (const auto & name: d->getMatrixView()->getRowNames()) {
+                result.emplace_back(PathElement(i) + name);
+            }
+
+        }
+        return result;
+    }
+
+    virtual vector<RowHash>
+    getRowHashes(ssize_t start = 0, ssize_t limit = -1) const
+    {
+        std::vector<RowHash> result;
+        for (const auto & it: rowIndex) {
+            result.emplace_back(it.first);
+        }
+        return result;
+    }
+
+    virtual bool knownRow(const Path & rowName) const
+    {
+        int idx = getIdxFromRowName(rowName);
+        if (idx == -1) {
+            return false;
+        }
+        return datasets[idx]->getMatrixView()->knownRow(rowName.tail());
+    }
+
+    virtual bool knownRowHash(const RowHash & rowHash) const
+    {
+        // Unused ?
+        return rowIndex.find(rowHash) != rowIndex.end();
+        //return rowIndex.getDefault(rowHash, 0) != 0;
+    }
+
+    virtual RowName getRowName(const RowHash & rowHash) const
+    {
+        const auto & it = rowIndex.find(rowHash);
+        if (it == rowIndex.end()) {
+            throw ML::Exception("Row not known");
+        }
+        const auto & idxAndHash = it->second;
+        return datasets[idxAndHash.first]->getMatrixView()->getRowName(idxAndHash.second);
+    }
+
+    // DEPRECATED
+    virtual MatrixNamedRow getRow(const RowName & rowName) const
+    {
+        throw ML::Exception("Unimplemented %s : %d", __FILE__, __LINE__);
+    }
+
+    virtual bool knownColumn(const Path & column) const
+    {
+        for (const auto & d: datasets) {
+            if (d->getMatrixView()->knownColumn(column)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    virtual ColumnName getColumnName(ColumnHash columnHash) const
+    {
+        for (const auto & d: datasets) {
+            try {
+                return d->getMatrixView()->getColumnName(columnHash);
+            }
+            catch (const ML::Exception & exc) {
```

Comment:
!!! we shouldn't ever have a try/catch unconditional like that.  Instead, we should use knownColumn